src/tests/t_kdb.py

```diff
@@ -336,6 +336,31 @@ def ldap_add(dn, objectclass, attrs=[]):
 realm.kinit('canon', password('canon'))
 realm.kinit('alias', password('canon'), ['-C'])
 
+# Test password history
+def test_pwhist(nhist):
+    output('Testing password history of size ' + `nhist`)
+    princ = 'pwhistprinc' + `nhist`
+    pol = 'pwhistpol'+`nhist`
+    realm.run([kadminl, 'addpol', '-history', `nhist`, pol])
+    realm.run([kadminl, 'addprinc', '-policy', pol, '-nokey', princ])
+    # Test nhist number of passwords and make sure that all previous passwords
+    # fail when trying to be changed to them.
+    cpw = lambda x,y: realm.run([kadminl, 'cpw', '-pw', `x`, princ],
+                                expected_code=y)
+
+    # Creates a combination of password attempts, with (x, 0) and ([0..x], 1)
+    cmb = lambda x: izip(chain([x], islice(count(), x+1)),
+                             chain([0], repeat(1)))
+    # Tries passwords, with one password change that should succeed followed by
+    # all previous passwords that should fail
+    # (0, 0) (0, 1) (1, 0) (0, 1), (1, 1)...
+    list(starmap(cpw, chain.from_iterable(imap(cmb, islice(count(), nhist)))))
+
+    # Test that keys are rotated out correctly
+    cpw(nhist, 0)
+    cpw(0, 0)
+map(test_pwhist, imap(pow, repeat(2), range(5)))
```

Comment:
These tests run hundreds of commands, way more than is necessary to hit the edge cases of password history.  I also think it's hard to read this testing code because it overuses itertools functions, although I expect that to be a topic of some debate.